typeahead_helper.js: Order streams by activity in typeahead.

cory2067 opened the pull request:
The first order sort of streams is still based on whether a match is found in the name, start of description, middle of description, etc.
But now the second order sort is by activity -- more specifically, the same exact order found in the left sidebar. This means pinned streams first, active streams, then inactive streams.

Fixes #4508

smarx commented:
*Automated message from Dropbox CLA bot*

@cory2067, it looks like you've already signed the Dropbox CLA. Thanks!

cory2067 commented:
I wonder if sorting _exactly_ like the sidebar is the best solution. So, the sidebar arranges streams in the order of: pinned, active, inactive.
But within each of those categories, it orders the streams alphabetically. 
Do you think alphabetical is the most logical sub-ordering? Perhaps it could be better to sort these by subscriber count, or something.

timabbott commented on static/js/typeahead_helper.js:
```diff
@@ -207,17 +207,27 @@ exports.compare_by_sub_count = function (stream_a, stream_b) {
+// Find which stream comes first in the sidebar listing order
+exports.compare_by_sidebar = function (stream_a, stream_b) {
+    var a_score = stream_data.is_active(stream_a.name) + stream_a.pin_to_top * 2;
+    var b_score = stream_data.is_active(stream_b.name) + stream_b.pin_to_top * 2;
```
For readability, I'd probably spell this `2 * stream_a.pin_to_top + stream_data.is_active(stream_a.name)`

That said, I'm not a fan of implicitly treating bools and ints in this sort of context, so it might be better to extract a `score` function.

(As a sidenote, it's theoretically possible to reuse code from `stream_sort.js`, but I'm not sure it's worth it.).

cory2067 replied:
By score function, would you mean something like:
`function score (stream) {
    if (stream.pin_to_top) {
        return 2;
    }
    if (stream_data.is_active(stream.name)) {
        return 1;
    } 
    return 0;
}`
It's quite a bit more bulky, but would you prefer something like that to a solution implicity treating bools as ints?

timabbott replied:
yep.  The bulk isn't a material cost -- in a large project, code clarity is far more important than brevity.

I'd use `stream_score += 2`, though, BTW; it's a little safer against someone adding a new rule at the top of the group :)

cory2067 commented:
Thanks for the tips on code clarity -- I'll definitely keep things like that in mind next time.
I added a unit test for this, and changed it a bit so the sub-ordering is done by subscriber count.

showell commented on static/js/typeahead_helper.js:
```diff
@@ -207,17 +207,38 @@ exports.compare_by_sub_count = function (stream_a, stream_b) {
+exports.compare_by_sidebar = function (stream_a, stream_b) {
+    var a_score = activity_score(stream_a);
+    var b_score = activity_score(stream_b);
+    if (a_score === b_score) {
```
This code is fine, but it's usually a good convention in these compare functions to early-return when the first criteria are unequal.

~~~ js
var diff = a_score - b_score

if (diff !== 0) {
    return diff;
}

return exports.sub_by_count(stream_a, stream_b);

showell commented:
There is now a method called `stream_sort.get_streams()` that returns all the sidebar streams in order, so I really think we should try to take advantage of that in the typeahead code, rather than duplicating the sort logic.

cory2067 commented:
I actually originally had my code base the ordering off that get_streams() function. I ended up changing it because I figured it would allow for more flexibility (for example, this code currently takes into account sub count, which is not possible with get_streams()). But if that isn't so important, I could revert my old code.

showell commented:
Huh, good point, I'm fine with a custom sort.

cory2067 commented:
amended function to early-return when scores are different

showell commented on static/js/typeahead_helper.js:
```diff
@@ -207,17 +207,37 @@ exports.compare_by_sub_count = function (stream_a, stream_b) {
+function activity_score(stream) {
+    var stream_score = 0;
+    if (stream.pin_to_top) {
+        stream_score += 2;
```
We can simplify this function to just explicitly return 2, 1, and 0 for the scores.  No need for `+=`.

cory2067 replied:
The idea with the `+=` was that it could possibly return 3 to prioritize active pinned streams over inactive pinned streams. But if this extra level of sorting isn't useful, I can go ahead and remove it.

showell replied:
Oh, ok, I think four tiers makes more sense, and we should just make the code more explicit on that.

cory2067 commented:
Added comments to make intention of code more clear.

showell commented:
Does this change make `compare_by_sub_count` obsolete?  If so, that could be a second clean-up commit.

cory2067 commented:
Yeah, it looks like `compare_by_sub_count` isn't used anywhere else, so will clean up.

timabbott commented:
@cory2067 you should extend the tests to cover the new subscriber count login, and maybe also add a final alphabetical tiebreak if the subscriber counts are the same?  

Otherwise, lgtm.

cory2067 commented:
Ok, that should be all set.

timabbott commented:
This looks great!  Since the commits had a lot of churn in this branch, I squashed and merged as 088d88115988eee233d3873caf85d189bb493c32, which reads as pretty clean.

Thanks @cory2067!  I suspect a lot of folks will enjoy this :).
